File: download.jl
# GloVe on text8 — Gideon (Julia)
# Run as script:  julia --project=. download.jl
# Or in REPL:     include("download.jl")  then call main()

using Downloads, SparseArrays, Random, LinearAlgebra, Printf
using Gideon

function ensure_text8(path)
    if !isfile(path)
        println("Downloading text8...")
        Downloads.download("http://mattmahoney.net/dc/text8.zip", path * ".zip")
        run(`unzip -o $(path * ".zip") -d $(dirname(path))`)
    end
end

function build_vocab(tokens; min_count = 5)
    counts = Dict{String, Int}()
    for t in tokens
        counts[t] = get(counts, t, 0) + 1
    end
    vocab = sort([w for (w, c) in counts if c >= min_count])
    word_to_id = Dict(w => i for (i, w) in enumerate(vocab))
    return vocab, word_to_id
end

function build_tcm(tokens, word_to_id; window = 5)
    ids = [get(word_to_id, t, 0) for t in tokens]
    V   = length(word_to_id)
    I = Int[]; J = Int[]; W = Float64[]
    for c in eachindex(ids)
        i = ids[c]; i == 0 && continue
        for d in 1:window
            c + d > length(ids) && break
            j = ids[c + d]; j == 0 && continue
            push!(I, i, j); push!(J, j, i); push!(W, 1/d, 1/d)
        end
    end
    return sparse(I, J, W, V, V)
end

function nearest(word, E, word_to_id, id_to_word; k = 5)
    id = get(word_to_id, word, 0)
    id == 0 && return String[]
    v    = E[:, id]
    sims = [dot(E[:, j], v) / (norm(E[:, j]) * norm(v) + 1e-12) for j in axes(E, 2)]
    sims[id] = -Inf
    return id_to_word[sortperm(sims, rev = true)[1:k]]
end

# Julia REPL — after include("download.jl"); res = main()
function analogy(a, b, c, E, word_to_id, id_to_word; k = 5)
    ids = [get(word_to_id, w, 0) for w in (a, b, c)]
    any(==(0), ids) && return String[]
    v = E[:, ids[2]] - E[:, ids[1]] + E[:, ids[3]]
    sims = [dot(E[:, j], v) / (norm(E[:, j]) * norm(v) + 1e-12) for j in axes(E, 2)]
    foreach(i -> sims[i] = -Inf, ids)
    id_to_word[sortperm(sims, rev = true)[1:k]]
end

# Julia
function wordsim_spearman(pairs, human_scores, E, word_to_id)
    model_sims = Float64[]
    valid_human = Float64[]
    for (i, (w1, w2)) in enumerate(pairs)
        i1, i2 = get(word_to_id, w1, 0), get(word_to_id, w2, 0)
        (i1 == 0 || i2 == 0) && continue
        v1, v2 = E[:, i1], E[:, i2]
        push!(model_sims, dot(v1, v2) / (norm(v1) * norm(v2) + 1e-12))
        push!(valid_human, human_scores[i])
    end
    # Spearman = Pearson on ranks
    r1 = invperm(sortperm(model_sims)) .|> Float64
    r2 = invperm(sortperm(valid_human)) .|> Float64
    dot(r1 .- mean(r1), r2 .- mean(r2)) / (std(r1) * std(r2) * (length(r1) - 1))
end

function main()
    text8 = expanduser("~/text8")
    ensure_text8(text8)

    wiki = open(readline, text8)
    tokens = split(wiki)

    id_to_word, word_to_id = build_vocab(tokens)
    C = build_tcm(tokens, word_to_id)
    @printf "vocab: %d words  |  co-occ nnz: %d\n" length(word_to_id) nnz(C)

    model = GloVe(rank = 50, x_max = 10.0, learning_rate = 0.15, verbose = true)
    fit!(model, C; n_iter = 20, rng = MersenneTwister(42))
    E = get_embeddings(model)   # rank × vocab

    println("\n── Nearest neighbors ──────────────────────────────────────────")
    for w in ("paris", "london", "king", "computer")
        haskey(word_to_id, w) || continue
        @printf "%-10s  %s\n" w join(nearest(w, E, word_to_id, id_to_word), "  ")
    end

    println("\n── Analogies ────────────────────────────────────────────────")
    analogy("man", "king", "woman", E, word_to_id, id_to_word)   # expect: queen
    analogy("paris", "france", "berlin", E, word_to_id, id_to_word)  # expect: germany

    return (; model, E, word_to_id, id_to_word, C)
end

main()

File: download.r
# GloVe on text8 — text2vec (R)
# Run as script:  Rscript download.r
# Or in REPL:     source("download.r")  then call main()

suppressPackageStartupMessages({
  library(text2vec)
  library(parallel)
})

ensure_text8 <- function(path) {
  if (!file.exists(path)) {
    message("Downloading text8...")
    download.file("http://mattmahoney.net/dc/text8.zip", paste0(path, ".zip"), mode = "wb")
    unzip(paste0(path, ".zip"), exdir = dirname(path))
  }
}

build_tcm <- function(tokens, min_count = 5L, window = 5L) {
  it    <- itoken(tokens, progressbar = FALSE)
  vocab <- prune_vocabulary(create_vocabulary(it), term_count_min = min_count)
  vect  <- vocab_vectorizer(vocab)
  tcm   <- create_tcm(it, vect, skip_grams_window = window)
  list(tcm = tcm, vocab = vocab)
}

train_glove <- function(tcm, rank = 50L, x_max = 10, learning_rate = 0.15,
                        n_iter = 20L, n_threads = detectCores()) {
  glove <- GlobalVectors$new(rank = rank, x_max = x_max, learning_rate = learning_rate)
  wv    <- glove$fit_transform(tcm, n_iter = n_iter, convergence_tol = 0.01,
                               n_threads = n_threads)
  E <- wv + t(glove$components)   # vocab x rank  (rows = words)
  list(glove = glove, E = E)
}

nearest <- function(word, E, k = 5L) {
  v    <- E[word, , drop = FALSE]
  sims <- sim2(E, v, method = "cosine", norm = "l2")[, 1]
  head(sort(sims[names(sims) != word], decreasing = TRUE), k)
}

analogy <- function(a, b, c, k = 5) {
  v    <- E[b, ] - E[a, ] + E[c, ]
  sims <- (E / sqrt(rowSums(E^2))) %*% (v / sqrt(sum(v^2)))
  top  <- order(sims, decreasing = TRUE)
  top  <- top[!rownames(E)[top] %in% c(a, b, c)][1:k]
  setNames(sims[top], rownames(E)[top])
}

main <- function() {
  text8 <- path.expand("~/text8")
  ensure_text8(text8)

  tokens <- space_tokenizer(readLines(text8, n = 1L, warn = FALSE))

  res   <- build_tcm(tokens)
  cat(sprintf("vocab: %d words  |  co-occ nnz: %d\n", nrow(res$vocab), Matrix::nnzero(res$tcm)))

  out   <- train_glove(res$tcm)
  E     <<- out$E
  glove <<- out$glove

  cat("\n── Nearest neighbors ──────────────────────────────────────────\n")
  for (w in c("paris", "london", "king", "computer")) {
    if (w %in% rownames(E))
      cat(sprintf("%-10s  %s\n", w, paste(names(nearest(w, E)), collapse = "  ")))
  }

  cat("\n── Analogies ────────────────────────────────────────────────\n")
  analogy("man", "king", "woman", 10)
  analogy("paris", "france", "berlin", 10)

  invisible(list(glove = glove, E = E, vocab = res$vocab, tcm = res$tcm))
}

main()
